hal/src/thumbv7em/clock/v2.rs

```diff
@@ -1,98 +1,859 @@
+//!         xosc::Xosc,
+//!     },
+//!     gpio::Pins,
+//!     pac::Peripherals,
+//!     time::U32Ext,
+//! };
+//! ```
+//!
+//! To create an instance of [`Xosc`], we will first need to identify which of
+//! the two XOSC clocks we will use. Suppose an external crystal is attached to
+//! pins `PA14` and `PA15`. These pins feed the XOSC0 clock, so we will want to
+//! create an instance of [`Xosc0`]. Note that `Xosc0<M>` is merely an alias for
+//! `Xosc<Xosc0Id, M>`. Here, [`Xosc0Id`] represents the
+//! [*identity*](self#id-types) of the XOSC0 clock, rather than the clock
+//! itself, and `M` represents the XOSC [`Mode`](xosc::Mode).
+//!
+//! Next, we access the [`Tokens`] struct to extract the corresponding
+//! [`XoscToken`] for XOSC0, and we trade the PAC `PORT` struct for the
+//! [`gpio::Pins`] struct to access the GPIO pins. We can then call
+//! [`Xosc::from_crystal`] to trade the token and [`Pin`]s to yield an instance
+//! of [`Xosc0`]. In doing so, we also provide the oscillator frequency.
+//!
+//! Finally, we can chain a call to the [`Xosc::enable`] method to enable the
+//! XOSC and return an instance of [`EnabledXosc0<M, N>`], which is simply an
+//! alias for `Enabled<Xosc0<M>, N>`. In this case, we get
+//! `EnabledXosc0<CrystalMode, U0>`.
+//!
+//! ```no_run
+//! # use atsamd_hal::{
+//! #     clock::v2::{
+//! #         clock_system_at_reset,
+//! #         xosc::Xosc,
+//! #     },
+//! #     gpio::Pins,
+//! #     pac::Peripherals,
+//! #     time::U32Ext,
+//! # };
+//! # let mut pac = Peripherals::take().unwrap();
+//! # let (buses, clocks, tokens) = clock_system_at_reset(
+//! #     pac.OSCCTRL,
+//! #     pac.OSC32KCTRL,
+//! #     pac.GCLK,
+//! #     pac.MCLK,
+//! #     &mut pac.NVMCTRL,
+//! # );
+//! let pins = Pins::new(pac.PORT);
+//! let xosc0 = Xosc::from_crystal(
+//!     tokens.xosc0,
+//!     pins.pa14,
+//!     pins.pa15,
+//!     8.mhz(),
+//! ).enable();
+//! ```
+//!
+//! Next, we want to use a DPLL to multiply the 8 MHz crystal clock up to 100
+//! MHz. Once again, we need to decide between two instances of a clock, because
+//! each chip has two [`Dpll`]s. This time, however, our decision between
+//! [`Dpll0`] and [`Dpll1`] is arbitrary.
+//!
+//! Also note that, like before, `Dpll0<I>` and `Dpll1<I>` are aliases for
+//! `Dpll<Dpll0Id, I>` and `Dpll<Dpll1Id, I>`. [`Dpll0Id`] and [`Dpll1Id`]
+//! represent the *identity* of the respective DPLL, while `I` represents the
+//! [`Id` type](self#id-types) for the [`Source`] driving the DPLL. In this
+//! particular case, we aim to create an instance of `Dpll0<Xosc0Id>`.
 //!
-//! This module provides the set of abstractions allowing a user to manipulate
-//! an ATSAMD's clocking system in a safe and expressive manner.
+//! Only certain clocks can drive the DPLL, so `I` is constrained by the
+//! [`DpllSourceId`] trait. Specifically, only the [`Xosc0Id`], [`Xosc1Id`],
+//! [`Xosc32kId`] and [`GclkId`] types implement this trait.
```

Comment:
`GclkId` is not really a type, but I guess that is sort of nitpicking and more precision would muddy the waters in a sentence. So I assume it is fine the way it is